scripts/train_agents.py

```diff
@@ -2,16 +2,200 @@
+        args.exp_dir = root_adv + '/' + main_agent_type + '-' + adversary_type + 'play' + '/' + str(round)
+        PwADVs( args=args, 
+                agent_folder_path = root_adv + '/' + main_agent_type + '-' + adversary_type + 'play' + '/' + str(round-1), 
+                agent_file_tag = pwadv_tag + '/' + CheckedPoints.FINAL_TRAINED_MODEL,
+                adv_folder_paths = [root_adv + '/' + adversary_type + '/' + str(round)], 
+                adv_file_tag = adv_tag + '/' + checked_adversary)
+        
+def MultiAdversaryPlay(args, 
+                        exp_tag = 'M2FP', 
+                        main_agent_path = None,
+                        main_agent_type = LearnerType.SUPPORTER, 
+                        adversary_type = LearnerType.SELFISHER, 
+                        checked_adversary = CheckedPoints.FINAL_TRAINED_MODEL, 
+                        how_long_init = 4.0,
+                        how_long_for_agent = 4.0,
+                        how_long_for_adv = 4.0,
+                        rounds_of_advplay = 401,
+                        reward_magnifier = 3.0):
+    
+    args.dynamic_reward = False
+    args.final_sparse_r_ratio = 0.5
+    if main_agent_path is None:
+        how_long = how_long_init
+        set_input(args=args, quick_test=quick_test, how_long=how_long)
+        args.SP_seed = args.ADV_seed
+        args.SP_h_dim = args.ADV_h_dim
+        args.learner_type = main_agent_type
+        args.reward_magnifier = reward_magnifier
+        args.exp_dir = exp_tag 
+        SP(args=args, pop_force_training=True)
+        sp_tag = 'sp_s' + args.SP_seed + '_h' + args.SP_h_dim + '_tr(SP)_ran'
+        main_agent_path = exp_tag + '/' + sp_tag
+    root = 'agent_models/' + main_agent_path
+    root_adv = root + '/' + exp_tag
+
+    adv_tag = 'adv_s' + args.ADV_seed + '_h' + args.ADV_h_dim + '_tr(H)_ran'
+    pwadv_tag = 'pwadv_s' + args.ADV_seed + '_h' + args.ADV_h_dim + '_tr(SP_SPADV)_ran'
+
+    how_long = how_long_for_adv
+    set_input(args=args, quick_test=quick_test, how_long=how_long)
+    args.learner_type = adversary_type
+    args.reward_magnifier = reward_magnifier
+    args.exp_dir = root_adv + '/' + adversary_type + '/0'
+    ADV(args=args, 
+        agent_folder_path = root, 
+        agent_file_tag= CheckedPoints.BEST_EVAL_REWARD)
+
+    how_long = how_long_init + how_long_for_agent
+    set_input(args=args, quick_test=quick_test, how_long=how_long)
+    args.learner_type = main_agent_type
+    args.reward_magnifier = reward_magnifier
+    args.exp_dir = root_adv + '/' + main_agent_type + '-' + adversary_type + 'play' + '/0'
+    PwADVs( args=args, 
+            agent_folder_path = root, 
+            agent_file_tag = CheckedPoints.BEST_EVAL_REWARD,
+            adv_folder_paths = [root_adv + '/' + adversary_type + '/0'], 
+            adv_file_tag = adv_tag + '/' + checked_adversary)
+    ###################################################################
+    for round in range(1,rounds_of_advplay):
```

Comment:
These for loops should be moved into `train_helper.py`

Reply:
I prefer this loop to be in train_agents.py

My point is that when we ran a round of training process, and a round of other training process, we usually clearly state it in train_agents.py. Here, we just  run multiple rounds of two methods iteratively. Therefore, if we follow our tradition, then we may want to do the same and keep them in train_agents.py

Reply:
The way I see` train_agents.py` is that we state all the _variables_ that are required to train using the specific method (i.e. N-1-SP, FCP, etc) and then call an function from `train_helper.py` and pass all those _variables_ and we blindly expect that that function will take care of the training the way we want it. 

Right now, your logic for the way SingleAdversaryPlay/MultiAdversaryPlay is trained (such as the for loop) is in train_agents.py. I think ideally, there should be a function in train_agents.py, such as this:

```
def ADVERSARY_PLAY(args, force_training):
# set all the required variables here such as ...
    args.dynamic_reward
    args.final_sparse_r_ratio
    adversary_type 
    main_agent_path
    total_training_time_for_agent
    total_training_time_for_adv
    rounds_of_adv_play
    reward_magnifier
    type_of_adversary_play = # etiher multi adversary play or single adversary play

# After setting all these variables call the function that will take care of the training for you.
   get_agent_trained_with_adversary_play(all the variables above)
```

Then in `train_helper.py` you will define `get_agent_trained_with_adversary_play`.

This way, the code will be a lot more readable. Whenever we want to train an adversary agent, we can just set the all the variables inside `ADVERSARY_PLAY` in `train_agents.py` and don't have to worry about how it's going to be trained or where it's going to be saved, because it is all being taken care of in `train_helper.py`.